addons/base_geolocalize/models/base_geocoder.py

```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import json
+import urllib2
+import logging
+
+from odoo import api, models, tools, _
+from odoo.exceptions import UserError
+
+
+_logger = logging.getLogger(__name__)
+
+
+class GeoCoder(models.AbstractModel):
+    """
+    Abstract class used to call Geolocalization API and convert addresses
+    into GPS coordinates.
+    """
+    _name = "base.geocoder"
+
+    @api.model
+    def geo_query_address(self, street=None, zip=None, city=None, state=None,
+                          country=None):
+        """ Converts address fields into a valid string for querying
+        geolocation APIs.
+        :param street: street address
+        :param zip: zip code
+        :param city: city
+        :param state: state
+        :param country: country
+        :return: formatted string
+        """
+        provider = self.env['ir.config_parameter'].get_param(
+            'base_geolocalize.provider', 'openstreetmap')
+        if hasattr(self, '_geo_query_address_' + provider):
+            # Makes the transformation defined for provider
+            service = getattr(self, '_geo_query_address_' + provider)
+            return service(street, zip, city, state, country)
+        else:
+            # By default, join the non-empty parameters
+            return tools.ustr(', '.join(filter(None, [
+                street,
+                ("%s %s" % (zip or '', city or '')).strip(),
+                state,
+                country])))
+
+    @api.model
+    def geo_find(self, addr):
+        """Use a location provider API to convert an address string into
+        a latitude, longitude tuple.
+        Here we use Openstreetmap Nominatim by default.
+        :param addr: Address string passed to API
+        :return: (latitude, longitude) or None if not found
+        """
+        provider = self.env['ir.config_parameter'].get_param(
+            'base_geolocalize.provider', 'openstreetmap')
+        if not hasattr(self, '_call_' + provider):
+            raise UserError(_(
+                'Provider %s is not implemented for geolocation service.')
+                % provider
+            )
+        try:
+            service = getattr(self, '_call_' + provider)
+            result = service(addr)
+            return result
+        except UserError:
```

Comment:
maybe try catching an AttributeError in case the user-provided value is wrong and display a meaningful error

Reply:
this is already tested at line 57, I don't think any invalid user value can come here. The address parameter is given by the geo_query_address method.

Reply:
@ecino The testing on line 57 is wrong IMO and should be in the try block.  Something like..

Why not just 
```python
# Remove hasattr
try:
   service = getattr(...)
    result = ...
except AttributeError:
    raise UserError(...)
except UserError(...)
    raise
except:
   _logger = ...
   result = None
return result
```

Reply:
Indeed, I missed the line 57. However, as said @gdgellatly it is better to try and catch the possible errors (especially that in most cases, it will work just fine).
cf [Asking for permission instead of forgiveness](https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html)